prereise/gather/griddata/hifld/data_process/transmission.py

```diff
@@ -27,6 +29,152 @@ def check_for_location_conflicts(substations):
         )
 
 
+def map_lines_to_substations_using_coords(substations, lines, **kwargs):
+    """Map lines to substations using coordinates
+
+    :param pandas.DataFrame substations: data frame of substations.
+    :param pandas.DataFrame lines: data frame of all lines.
+    :param \\*\\*kwargs: optional arguments.
+    :return: (*pandas.DataFrame*) -- indices: lines id, columns: lines
+        endpoint coordinates (*FROM* and *TO*) and id of substations found
+        at these coordinates (*FROM_SUB* and *TO_SUB*)
+    """
+    rounding = 3 if "rounding" not in kwargs else kwargs["rounding"]
+    subcoord2subid = (
+        substations.round(rounding).groupby(["LONGITUDE", "LATITUDE"]).groups
+    )
+    print(
+        f"initial number of substations: {substations.shape[0]}\n"
+        f"substations with same location after rounding: "
+        f"{substations.shape[0] - len(subcoord2subid)}\n"
+    )
+
+    lines_coord_from = lines["COORDINATES"].map(
+        lambda x: (round(x[0][0], rounding), round(x[0][1], rounding))
+    )
+    lines_coord_to = lines["COORDINATES"].map(
+        lambda x: (round(x[-1][0], rounding), round(x[-1][1], rounding))
+    )
+    line2sub = lines_coord_from.to_frame(name="FROM").join(
+        lines_coord_to.to_frame(name="TO")
+    )
+
+    for endpoint in ["FROM", "TO"]:
+        line2sub[f"{endpoint}_SUB"] = line2sub[endpoint].apply(
+            lambda x: list(subcoord2subid[x]) if x in subcoord2subid else np.NaN
```

Comment:
If converting to `list` is not essential, we could do `.apply(lambda x: subcoord2subid.get(x, np.NaN)` here.

Reply:
Done